### ISSUE TITLE
Catch when callbacks are missing a return.

### ISSUE THREAD

**Issue by sustained** [microsoft/TypeScript#36472]
## Search Terms

- array callback return
- eslint array-callback-return
- catch missing return for callbacks

## Suggestion

TypeScript does not recognise this as an error even though it clearly is - `findIndex` takes a callback which should return a boolean, returning `undefined` is IMO in 99% of cases a mistake.

```ts
[1,2,3].findIndex(e => {
  e === 1
})
```

I'm aware that ESLint can do this but not everyone uses ESLint, also ESLint can be a huge pain to configure sometimes, especially when Prettier, TypeScript and other things are involved.

## Use Cases

It will definitely catch errors, IMHO, for people who don't use ESLint or do not have the rule `array-callback-return`.

## Examples

No error:
```ts
[1,2,3].findIndex(e => 
  e === 1
)
```

No error:
```ts
[1,2,3].findIndex(e => e === 1)
```

A warning/errror: "Did you forget a return?"
```ts
[1,2,3].findIndex(e => {
  e === 1
})
```

I also think that if there is an *explicit return* of a falsey value such as `null`/`undefined` then that should be fine. It's the lack of any return at all which should trigger an error.

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by sustained** [comment]
Note that this should apply to anything which is provided with a testing function, not just `findIndex`.

**Comment by jcalz** [comment]
Until and unless this gets implemented, you might consider modifying the standard library typings in your local code base to get something close to this behavior:

```ts
interface Array<T> {
  findIndex2<B extends (void extends B ? boolean : unknown)>(
    cb: (v: T, i: number, arr: this) => B): number
}

[1, 2, 3].findIndex2(e => e === 1); // okay
[1, 2, 3].findIndex2(e => { e === 1 }); // error! void is not assignable to boolean
[1, 2, 3].findIndex2(e => { e === 1; return undefined; });// okay
[1, 2, 3].findIndex2(e => { e === 1; return; }); // error, void is not assignable to boolean 🤷‍♂️
```

[Playground link](https://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgIJSnAngHgCoB8yA3gLABQyyMoAJgJIi0QAeATDgELKuRMDOyABQA3APbBaPFn1qDuAfmQAjMWIA2EOCGQAuZAFcQAaxBiA7iACUBIRSpUEy-aP14ANMmD6QBgLbK0J5wGPpgABbA-FbIALxEnFY+-oFQFAC+FBQA2gCMnmyeAMwAugB0NEyMzOxCKPE8cbGxyLlWANzIAPRdyGLG2Dn5yIXIpRV01axsdXFExI3NLbnI6R3dvdBQYlAAhMjikl6CZmDIcPz8wADmIHDKmshgYipqmtpDBcXllQxM07MGgt6s1Wp0oBAwAYoDojMxKhBaJ01u0en0BlhPiNvhMqv9avV5otQblwZDoSBkes0Vsdp5DlIoshTudLjc7g8UM9XhotDpAHwbgHZdwCwBIAhMkA8H9Zcj2IA)

This is ugly enough that I wouldn't seriously suggest doing it upstream.  Less ugly is just requiring that the callback return `boolean`:

```ts
interface Array<T> {
  findIndex2(cb: (v: T, i: number, arr: this) => boolean): number
}

[1, 2, 3].findIndex2(e => e === 1); // okay
[1, 2, 3].findIndex2(e => { e === 1 }); // error! void is not assignable to boolean
[1, 2, 3].findIndex2(e => { e === 1; return undefined; });// error! undefined is not boolean
[1, 2, 3].findIndex2(e => { e === 1; return; }); // error, void is not assignable to boolean 
```
[Playground link](https://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgIJSnAngHgCoB8yA3gLABQyyMoAJgJIi0QAeATABQIBGAXMhwBu-PABpkwfiACuAW27RxcDPzAALYAGcAlMgC8RbgHsjAGwhwQ2qXIVQKAXwoUA2gEZxbcQGYAugDoaJkZmdg4UA2QIvT1kN20AbmQAemTkIwBrbFcPZC9kP0C6ENZOCKJiKP0YuOQHRJS06CgjKABCZEEjYFoJTWQQIzBkOE1NYABzEDhuc2QwI2RjMwsQHM8fAKCGJlLw-QqqmNi3JKgIMGkoEGRpXaCIWiT6hNSojFaOu+YH3q0BoYjMaTaazFALJYmcyWdZ5TZFYK7MLlEhHGqnZDnS7XZ4NN7NVriLo9PoA4ajcZTGZzCHLaE3ZzkChUIA)

which also wouldn't be suitable for upstream because it would likely break a lot of existing code that relies on truthy/falsy returns.

**Comment by sustained** [comment]
I had considered extending `lib.d.ts` in the discussion on Discord, actually.

The issue with requiring it to return `boolean`, less ugly be that as it may, is that it could break a lot of existing code.

Do you think that it would make sense for this to be the behaviour when `noImplicitReturns` is enabled?

![image](https://user-images.githubusercontent.com/5995080/73273933-a2503f00-41dc-11ea-820d-e56750f5db1c.png)

**Comment by RyanCavanaugh** [comment]
We'd need some way to identify that this particular kind of function is "expecting" an explicit `return` value. The array predicates are generally typed to return `unknown` since people will use pretty much anything as a truthy/falsy test, so this appears to be a valid test that happens to return the falsy value `undefined`.

**Comment by falsandtru** [comment]
Listing all the types without void would resolve this as follows.

```ts
interface Array<T> {
    findIndex(predicate: (value: T, index: number, obj: T[]) => undefined | boolean | number | bigint | string | symbol | null | object, thisArg?: any): number;
}
```
